examples/full_pipeline.rs

```diff
@@ -227,74 +168,43 @@ impl JoinSumOperator {
     }
 }
 
-struct JoinSumOperatorState {
-    sum: usize,
-    current_timestamp: Timestamp,
-}
-
-#[allow(dead_code)]
-impl JoinSumOperatorState {
-    fn new() -> Self {
-        Self {
-            sum: 0,
-            current_timestamp: Timestamp::Bottom,
-        }
-    }
-
-    fn add(&mut self, value: usize) {
-        self.sum += value;
-    }
-
-    fn get_sum(&self) -> usize {
-        self.sum
-    }
-}
-
-impl StateT for JoinSumOperatorState {
-    fn commit(&mut self, timestamp: &Timestamp) {
-        self.current_timestamp = timestamp.clone();
-    }
-
-    fn get_last_committed_timestamp(&self) -> Timestamp {
-        self.current_timestamp.clone()
-    }
-}
-
-impl TwoInOneOut<JoinSumOperatorState, usize, usize, usize> for JoinSumOperator {
+impl TwoInOneOut<TimeVersionedState<usize>, usize, usize, usize> for JoinSumOperator {
     fn on_left_data(
         &mut self,
-        ctx: &mut TwoInOneOutContext<JoinSumOperatorState, usize>,
+        ctx: &mut TwoInOneOutContext<TimeVersionedState<usize>, usize>,
         data: &usize,
     ) {
-        ctx.get_state().add(*data);
-        let state = ctx.get_state().get_sum();
+        let current_timestamp = ctx.get_timestamp().clone();
+        let current_state = ctx.get_current_state().unwrap();
+        *current_state += *data;
 
         tracing::info!(
             "JoinSumOperator @ {:?}: Received {} on left stream, sum is {}",
-            ctx.get_timestamp(),
+            current_timestamp,
             data,
-            state
+            *current_state
         );
     }
 
     fn on_right_data(
         &mut self,
-        ctx: &mut TwoInOneOutContext<JoinSumOperatorState, usize>,
+        ctx: &mut TwoInOneOutContext<TimeVersionedState<usize>, usize>,
         data: &usize,
     ) {
-        ctx.get_state().add(*data);
-        let state = ctx.get_state().get_sum();
+        let current_timestamp = ctx.get_timestamp().clone();
+        let current_state = ctx.get_current_state().unwrap();
+        *current_state += *data;
 
         tracing::info!(
             "JoinSumOperator @ {:?}: Received {} on right stream, sum is {}",
-            ctx.get_timestamp(),
+            current_timestamp,
             data,
-            state
+            *current_state
         );
     }
 
-    fn on_watermark(&mut self, ctx: &mut TwoInOneOutContext<JoinSumOperatorState, usize>) {
-        let state = ctx.get_state().get_sum();
+    fn on_watermark(&mut self, ctx: &mut TwoInOneOutContext<TimeVersionedState<usize>, usize>) {
+        let state = *ctx.get_current_state().unwrap();
```

Comment:
```suggestion
        let state = ctx.get_current_state().unwrap();
```

Reply:
Similar problems with multiple mutable borrows if we do this.